systemd/watch.go

```diff
@@ -0,0 +1,155 @@
+package systemd
+
+import (
+	"context"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+// If a provider implements the Updater interface we will send pings when configmaps and secrets are updated.
+type Updater interface {
+	UpdateConfigMap(context.Context, *corev1.Pod, *corev1.ConfigMap) error
+	UpdateSecret(context.Context, *corev1.Pod, *corev1.Secret) error
+}
+
+type Watcher struct {
+	mu        sync.RWMutex
+	clientset *kubernetes.Clientset
+	// maps are indexed by "namespace.name"
+	configs map[string][]*corev1.Pod
+	secrets map[string][]*corev1.Pod
+}
+
+func newWatcher(clientset *kubernetes.Clientset) *Watcher {
+	return &Watcher{
+		configs:   make(map[string][]*corev1.Pod),
+		secrets:   make(map[string][]*corev1.Pod),
+		clientset: clientset,
+	}
+}
+
+func (w *Watcher) Run(p Updater) error {
+	cmwatcher, err := w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	swatcher, err := w.clientset.CoreV1().Secrets(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case event, ok := <-cmwatcher.ResultChan():
+			if !ok {
+				cmwatcher, _ = w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+				break
+			}
+			cm, ok := event.Object.(*corev1.ConfigMap)
+			if !ok {
+				continue
+			}
+			// event.Type == watch.Deleted ??
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				continue
+			}
+
+			namespace := cm.ObjectMeta.Namespace
+			name := cm.ObjectMeta.Name
+			w.mu.RLock()
+			pods := w.configs[namespace+"."+name]
+			w.mu.RUnlock()
+			if len(pods) != 0 {
+				klog.Infof("Configmap update %s/%s, notifying %s pods", namespace, name, len(pods))
+			}
+			for _, pod := range pods {
+				p.UpdateConfigMap(context.TODO(), pod, cm)
+			}
+		case event, ok := <-swatcher.ResultChan():
+			if !ok {
+				swatcher, _ = w.clientset.CoreV1().Secrets(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+				break
+			}
+			s, ok := event.Object.(*corev1.Secret)
+			if !ok {
+				continue
+			}
+			// event.Type == watch.Deleted ??
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				continue
+			}
+
+			namespace := s.ObjectMeta.Namespace
+			name := s.ObjectMeta.Name
+			w.mu.RLock()
+			pods := w.secrets[namespace+"."+name]
+			w.mu.RUnlock()
+			if len(pods) != 0 {
+				klog.Infof("Secret update %s/%s, notifying %s pods", namespace, name, len(pods))
+			}
+			for _, pod := range pods {
+				p.UpdateSecret(context.TODO(), pod, s)
+			}
+		}
+	}
+}
+
+// Watch wathces the config maps and secrets for the pod. It calls back into the
+// provider using the Updater interface methods.
+func (w *Watcher) Watch(pod *corev1.Pod) {
```

Comment:
Can you please add unit test to `Watch` and `Unwatch`? I'll cover integration/e2e tests later.

Reply:
done

Reply:
done